samples/azure-quantum/qae-numerical-integration/QuantumAmplitudeEstimation.ipynb

```diff
@@ -0,0 +1,1148 @@
+    "    for k in map(lambda x : 2 ** x, range(8)):\n",
+    "        spin_echo_res = RunQAESpinEcho.estimate_resources(theta=theta, k=k)\n",
+    "        res = RunQAE.estimate_resources(theta=theta, k=k)\n",
+    "\n",
+    "        def delta(key : str) -> int:\n",
+    "            return (res[key] - spin_echo_res[key])\n",
+    "\n",
+    "        delta_2q = str(np.round(delta('CNOT') / res['CNOT'] * 100, 2)) + '%'\n",
+    "        delta_1q = str(np.round((delta('QubitClifford') + delta('R') + delta('T')) / (res['QubitClifford'] + res['R'] + res['T']) * 100, 2)) + '%'\n",
+    "        deltas.append([k, delta_1q, delta_2q])\n",
+    "    df = pd.DataFrame(deltas, columns=[\"k\", \"1Q gate reduction\", \"2Q gate reduction\"])\n",
+    "    df = df.style.set_caption(f\"Differences with Spin Echo optimization at bit length {bitlen}\").hide_index()\n",
+    "    dfs.append(df)\n",
+    "\n",
+    "html_dfs = list(map(lambda df : df.to_html(), dfs))\n",
+    "HTML(\"<br>\".join(html_dfs))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We see that the Spin Echo circuit requires fewer gates across the board and for higher $k$ we get bigger benefits.\n",
+    "\n",
+    "## Maximum Likelihood Estimator (MLE)\n",
+    "\n",
+    "We will be using the MLE to get our optimal theta results with a modified optimization function that you can see in the code below as `loglikelihood`.\n",
+    "\n",
+    "To learn more about why and how this estimator works with QAE please see our other sample and [arXiv:1904.10246](https://arxiv.org/pdf/1904.10246.pdf).\n",
+    "\n",
+    "We use (and adapt) the MLE calculation methods from the work resulting from [arXiv:1904.10246](https://arxiv.org/pdf/1904.10246.pdf) that can be found on [GitHub](https://github.com/qiskit-community/qiskit-community-tutorials/blob/master/algorithms/SimpleIntegral_AEwoPE.ipynb)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 37,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def calculate_theta(hit_list, number, shots):\n",
+    "    \"\"\"\n",
+    "        calculate optimal theta values\n",
+    "            hit_list: list of count of obserbving \"1\" for qc_list\n",
+    "            numebr_grover_list: list of number of Grover operators        \n",
+    "            shots_list: list of number of shots\n",
+    "\n",
+    "        Return:\n",
+    "            theta_candidates: list of optimal theta\n",
+    "    \"\"\"\n",
+    "\n",
+    "    small = 1.e-15  # small valued parameter to avoid zero division\n",
+    "    confidence_level = 5  # confidence level to determine the search range\n",
+    "\n",
+    "    theta_candidates = []\n",
+    "    range_min = 0.0 + small\n",
+    "    range_max = 1.0 - small\n",
+    "    for igrover in range(len(number)):\n",
+    "\n",
+    "        def loglikelihood(p):\n",
+    "            ret = np.zeros_like(p)\n",
```

Comment:
If possible, vectorizing here instead of using an explicit for-loop will give you much more efficient and readable code.